File: Concurs Inot Java/Service.java
package me.service;

import me.events.EvenimentSchimbare;
import me.exceptii.ValidationException;
import me.model.*;
import me.observer.Observer;
import me.repository.IRepoAngajat;
import me.repository.IRepoParticipant;
import me.repository.IRepoProba;
import me.repository.IRepoProbaParticipant;
import me.validator.Validator;
import me.observer.Observable;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class Service implements IService{
    private final int defaultThreadNo = 10;
    private IRepoProba repoProba;
    private IRepoParticipant repoParticipant;
    private IRepoAngajat repoAngajat;
    private IRepoProbaParticipant repoProbaParticipant;
    private Validator<Proba> validatorProba;
    private Validator<Angajat> validatorAngajat;
    private Validator<Participant> validatorParticipant;
    //private List<Observer<EvenimentSchimbare>>  observers;
    private Map<String,IProbaObserver> clientiLogati;

    public Service(IRepoProba repoProba, IRepoParticipant repoParticipant,
                   IRepoAngajat repoAngajat, IRepoProbaParticipant repoProbaParticipant,
                   Validator<Proba> validatorProba, Validator<Angajat> validatorAngajat,
                   Validator<Participant> validatorParticipant) {
        this.repoProba = repoProba;
        this.repoParticipant = repoParticipant;
        this.repoAngajat = repoAngajat;
        this.repoProbaParticipant = repoProbaParticipant;
        this.validatorProba = validatorProba;
        this.validatorAngajat = validatorAngajat;
        this.validatorParticipant = validatorParticipant;
        this.clientiLogati = new ConcurrentHashMap<>();
       // observers = new ArrayList<>();
    }
    /**
       REPO PROBE
     */
    public Proba getProbaID(Integer id){ return repoProba.get(id); }

    public Proba saveProba(Integer id, float distanta, String stil) throws ValidationException, ServiceException {
        Proba p = new Proba(id,distanta, Stiluri.valueOf(stil));
        validatorProba.validate(p);
        return repoProba.save(p);
    }

    public Iterable<Proba> getAllProbe(){
        return repoProba.findAll();
    }

    /**
       REPO PARTICIPANTI
     */
    public Participant getParticipantID(Integer id){ return repoParticipant.get(id); }

    public synchronized Participant saveParticipant(Integer id, String nume, int varsta) throws ValidationException, ServiceException{
        Participant p = new Participant(id,nume,varsta);
        validatorParticipant.validate(p);
        return repoParticipant.save(p);
    }

    public synchronized List<Participant> findParticipantiDupaNume(String nume) throws ServiceException{
        return repoParticipant.findParticipantiDupaNume(nume);
    }

    /**
       REPO ANGAJATI
     */
    public Angajat saveAngajat(Integer id, String nume, int varsta, String username, String parola) throws ValidationException{
        Angajat a = new Angajat(id,nume,varsta,username,parola);
        validatorAngajat.validate(a);
        return repoAngajat.save(a);
    }

    public synchronized Angajat getAngajat(String username, String parola) throws ServiceException{
        return repoAngajat.getUsernameParola(username,parola);
    }

    /**
       REPO PROBE-PARTICIPANTI
     */
    public synchronized void saveProbaParticipant(ArrayList<Integer> probe, Integer idParticipant) throws ServiceException{
        probe.forEach(x->{
            repoProbaParticipant.save(new ProbaParticipant(x,idParticipant));
        });
        notifyAngajatiAdaugare();
    }

    public List<ProbaParticipant> listaProbeParticipantiDupaParticipant(Integer idParticipant){
        return repoProbaParticipant.findProbaParticipantDupaParticipant(idParticipant);
    }

    public List<ProbaParticipant> listaProbeParticipantiDupaProba(Integer idProba){
        return repoProbaParticipant.findProbaParticipantDupaProba(idProba);
    }

    public synchronized List<ProbaDTO> getProbeDTO() throws ServiceException{
        List<ProbaDTO> probeDTO = new ArrayList<>();
        ArrayList<Proba> probe = new ArrayList<>();
        repoProba.findAll().forEach(probe::add);
        probe.forEach(x->{
            probeDTO.add(new ProbaDTO(x.getId(),x.getDistanta(),x.getStil(),repoProbaParticipant.findProbaParticipantDupaProba(x.getId()).size()));
        });
        return probeDTO;
    }

    public synchronized List<ParticipantDTO> getParticipantiDTOcuProbaP(int p) throws ServiceException{
        List<ParticipantDTO> participantiDTO = new ArrayList<>();
        ArrayList<Participant> participanti = new ArrayList<>();
        repoProbaParticipant.findProbaParticipantDupaProba(p).forEach(x->{
            participanti.add(repoParticipant.get(x.getIdParticipant()));
        });
        participanti.forEach(x->{
            String listaProbe = "";
            for (ProbaParticipant probaParticipant : repoProbaParticipant.findProbaParticipantDupaParticipant(x.getId())) {
                listaProbe+=repoProba.get(probaParticipant.getIdProba());
                listaProbe+=", ";
            }
            participantiDTO.add(new ParticipantDTO(x.getId(),x.getNume(),x.getVarsta(),listaProbe));
        });
        return participantiDTO;
    }

    private void afiseazaClienti(){
        clientiLogati.forEach((x,y)->
                System.out.println(x));
    }

    public synchronized void login(Angajat angajat, IProbaObserver client) throws ServiceException{
        if(angajat!=null){
            if(clientiLogati.get(angajat.getUsername())!=null){
                throw new ServiceException("User deja logat!\n");
            }
            clientiLogati.put(angajat.getUsername(),client);
            //System.out.println("AICIIIIII!!!");
            //afiseazaClienti();
        }
        else throw new ServiceException("Autentificare esuata!\n");
    }

    public synchronized void logout(Angajat angajat, IProbaObserver client) throws ServiceException{
        IProbaObserver localClient = clientiLogati.remove(angajat.getUsername());
        if(localClient==null)
            throw new ServiceException("Angajatul" + angajat.getUsername()+ " nu e logat");

    }
    public synchronized void check(Angajat angajat) throws ServiceException{
        List<Angajat> angajatList = (List<Angajat>) repoAngajat.findAll();
        boolean gasit = false;
        for(Angajat a : angajatList){
            if(a.getUsername().equals(angajat.getUsername()) && a.getParola().equals(angajat.getParola())) {
                gasit = true;
                break;
            }
        }
        if(!gasit)
            throw new ServiceException("Datele introduse nu sunt inregistrate!");
    }

    private synchronized void notifyAngajatiAdaugare(){
        Iterator<Angajat> angajatList = repoAngajat.findAll().iterator();
        ExecutorService executor = Executors.newFixedThreadPool(defaultThreadNo);
        angajatList.forEachRemaining(x-> {
                    IProbaObserver client = clientiLogati.get(x.getUsername());
                    System.out.println(x.getUsername() + " " + client);
                    if (client != null) {
                        System.out.println("am intrat in client not null");
                        System.out.println("se notifica " + client.toString());
                        executor.execute(() -> {
                            try {
                                System.out.println("se notifica " + x.getUsername());
                                List<ProbaDTO> probeDTO = getProbeDTO();
                                client.updateProbeDTO(probeDTO);
                            } catch (ServiceException e) {
                                e.printStackTrace();
                            }
                        });
                    }
                });
        executor.shutdown();
    }

}

File: Concurs Inot C#/Service.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Problema_3_CSharp.me.validator;
using Problema_3_CSharp.me.repository;
using Problema_3_CSharp.me.model;

namespace Problema_3_CSharp.me.service
{
    public class Service :  IService
    {
        private IRepoAngajat repoAngajat;
        private IRepoProba repoProba;
        private IRepoParticipant repoParticipant;
        private IRepoProbaParticipant repoProbaParticipant;
        private IValidator<Angajat> validatorAngajat;
        private IValidator<Proba> validatorProba;
        private IValidator<Participant> validatorParticipant;
        private readonly IDictionary<string, IObserver> clientiLogati;

        public Service(IRepoAngajat repoAngajat, IRepoProba repoProba, IRepoParticipant 
            repoParticipant, IRepoProbaParticipant repoProbaParticipant,
            IValidator<Angajat> validatorAngajat, IValidator<Proba> validatorProba,
            IValidator<Participant> validatorParticipant)
        {
            this.repoAngajat = repoAngajat;
            this.repoProba = repoProba;
            this.repoParticipant = repoParticipant;
            this.repoProbaParticipant = repoProbaParticipant;
            this.validatorAngajat = validatorAngajat;
            this.validatorProba = validatorProba;
            this.validatorParticipant = validatorParticipant;
            clientiLogati = new Dictionary<string, IObserver>();
        }

        /*
         * REPO ANGAJAT
         */
        public Angajat SaveAngajat(int id, String nume, int varsta, String username, String parola)
        {
            lock (this)
            {
                Angajat a = new Angajat(id, nume, varsta, username, parola);
                validatorAngajat.Validate(a);
                return repoAngajat.Save(a);
            }
        }

        public Angajat GetAngajatUsernameParola(String username, String parola)
        {
            lock (this)
            {
                return repoAngajat.GetAngajatUsernameParola(username, parola);
            }
        }

        /*
         * REPO PROBA
         */
        public Proba SaveProba(int id, float distanta, String stil)
        {
            lock (this)
            {
                Proba p = new Proba(id, distanta, (Stiluri)Enum.Parse(typeof(Stiluri), stil));
                validatorProba.Validate(p);
                return repoProba.Save(p);
            }
        }

        public Proba GetProba(int id)
        {
            lock (this)
            {
                return repoProba.Get(id);
            }
        }

        /*
         * REPO PARTICIPANT
         */
        public Participant SaveParticipant(int id, String nume, int varsta)
        {
            lock (this)
            {
                Participant p = new Participant(id, nume, varsta);
                validatorParticipant.Validate(p);
                return repoParticipant.Save(p);
            }
        }

        public Participant GetParticipant(int id)
        {
            lock (this)
            {
                return repoParticipant.Get(id);
            }
        }

        public IEnumerable<Participant> GetParticipantiDupaNume(String nume)
        {
            lock (this)
            {
                return repoParticipant.FindParticipantiDupaNume(nume);
            }
        }

        /*
         * REPO PROBE-PARTICIPANTI
         */
        public void SaveProbaParticipant(List<int> probe, int idParticipant, int probaCurenta)
        {
            lock (this)
            {
                probe.ForEach(x =>
                {
                    repoProbaParticipant.Save(new ProbaParticipant(x, idParticipant));
                });
                NotifyAngajatiAdaugare(probaCurenta);
            }

        }

        public IEnumerable<ProbaParticipant> FindProbaParticipantDupaParticipant(int idParticipant)
        {
            lock (this)
            {
                return repoProbaParticipant.FindProbaParticipantDupaParticipant(idParticipant);
            }
        }

        public IEnumerable<ProbaParticipant> FindProbaParticipantDupaProba(int idProba)
        {
            lock (this)
            {
                return repoProbaParticipant.FindProbaParticipantDupaProba(idProba);
            }
        }

        public IEnumerable<ProbaDTO> GetProbeDTO()
        {
            Console.WriteLine("aici ajunge");
            lock (this)
            {
                Console.WriteLine("si aici ajunge");
                List<ProbaDTO> probeDTO = new List<ProbaDTO>();
                List<Proba> probe = new List<Proba>();
                repoProba.FindAll().ToList().ForEach(x => probe.Add(x));
                probe.ForEach(x =>
                {
                    probeDTO.Add(new ProbaDTO(x.Id, x.Distanta, x.Stil, repoProbaParticipant.FindProbaParticipantDupaProba(x.Id).Count()));
                });
                return probeDTO;
            }
        }

        public IEnumerable<ParticipantDTO> getParticipantiDTOCuProbaP(int p)
        {
            lock (this)
            {
                List<ParticipantDTO> participantiDTO = new List<ParticipantDTO>();
                List<Participant> participanti = new List<Participant>();
                repoProbaParticipant.FindProbaParticipantDupaProba(p).ToList().ForEach(x =>
                {
                    participanti.Add(repoParticipant.Get(x.IdParticipant));
                });
                participanti.ForEach(x =>
                {
                    String listaProbe = "";
                    repoProbaParticipant.FindProbaParticipantDupaParticipant(x.Id).ToList()
                    .ForEach(y =>
                    {
                        listaProbe += repoProba.Get(y.IdProba);
                        listaProbe += ", ";
                    });
                    participantiDTO.Add(new ParticipantDTO(x.Id, x.Nume, x.Varsta, listaProbe));
                });
                return participantiDTO;
            }
        }

        public void Login(Angajat a, IObserver client)
        {
            lock (this)
            {
                if (a != null)
                {
                    if (clientiLogati.ContainsKey(a.Username))
                        throw new ServiceException("User already logged!");
                    clientiLogati.Add(a.Username, client);
                }
                else throw new ServiceException("Authentication failes!");
            }
        }

        public void Logout(Angajat a, IObserver client)
        {
            IObserver localClient = clientiLogati[a.Username];
            if (localClient == null)
                throw new ServiceException("User " + a.Username + " is not logged in.");
            clientiLogati.Remove(a.Username);
        }

        public void Check(Angajat a)
        {
            List<Angajat> angajatList = (List<Angajat>)repoAngajat.FindAll();
            bool gasit = false;
            foreach (Angajat angajat in angajatList)
            {
                if (a.Username.Equals(angajat.Username) && a.Parola.Equals(angajat.Parola))
                {
                    gasit = true;
                    break;
                }
            }
            if (!gasit)
                throw new ServiceException("Datele introduse nu sunt inregistrate!");
        }

        private void NotifyAngajatiAdaugare(int probaCurenta)
        {
            foreach (var cl in clientiLogati.Keys)
            {
                IObserver Client = clientiLogati[cl];
                IList<ParticipantDTO> participants = getParticipantiDTOCuProbaP(probaCurenta).ToList();
                Task.Run(() => Client.Update(GetProbeDTO().ToList(), participants));
            }
        }
    }
}
